src/btree/bt_sync_obsolete.c

```diff
@@ -392,29 +426,40 @@ __checkpoint_cleanup_page_skip(
     }
 
     /*
-     * The checkpoint cleanup fast deletes the obsolete leaf page by marking it as deleted
-     * in the internal page. To achieve this,
+     * From this point, the page is on disk and checkpoint cleanup should NOT induce a read if at
+     * least of one of the following conditions is met:
      *
-     * 1. Checkpoint has to read all the internal pages that have obsolete leaf pages.
-     *    To limit the reading of number of internal pages, the aggregated stop durable timestamp
-     *    is checked except when the table is logged. Logged tables do not use timestamps.
+     * - The page is a leaf page with no overflow items because obsolete leaf pages with overflow
+     * keys/values cannot be fast deleted to free the overflow blocks. The overflow blocks can be
+     * freed during reconciliation after being marked dirty.
      *
-     * 2. Obsolete leaf pages with overflow keys/values cannot be fast deleted to free
-     *    the overflow blocks. Read the page into cache and mark it dirty to remove the
-     *    overflow blocks during reconciliation.
+     * - The page is not fully deleted (checked using the aggregated stop durable timestamp) and
+     *   - the table is not logged (logged tables do not use timestamps) or checkpoint cleanup is
+     * configured to reclaim space.
      *
      * FIXME: Read internal pages from non-logged tables when the remove/truncate
      * operation is performed using no timestamp.
      */
+    aggressive_mode = F_ISSET(S2C(session), WT_CONN_CKPT_CLEANUP_SKIP_INT);
+    logged = F_ISSET(S2BT(session), WT_BTREE_LOGGED);
+    page_fully_deleted = addr.ta.newest_stop_durable_ts != WT_TS_NONE;
+    if (addr.type == WT_ADDR_LEAF_NO || (!page_fully_deleted && (aggressive_mode || !logged))) {
+        /*
+         * While we may have decided to skip the page, we still want to read it if there is globally
+         * visible content.
+         */
+        if (!__sync_obsolete_can_process(session, addr.ta))
+            *skipp = true;
+    }
 
-    if (addr.type == WT_ADDR_LEAF_NO ||
-      (addr.ta.newest_stop_durable_ts == WT_TS_NONE &&
-        (F_ISSET(S2C(session), WT_CONN_CKPT_CLEANUP_SKIP_INT) ||
-          !F_ISSET(S2BT(session), WT_BTREE_LOGGED)))) {
+    if (*skipp) {
         __wt_verbose_debug2(
           session, WT_VERB_CHECKPOINT_CLEANUP, "%p: page walk skipped", (void *)ref);
         WT_STAT_CONN_DSRC_INCR(session, checkpoint_cleanup_pages_walk_skipped);
-        *skipp = true;
+    } else {
+        __wt_verbose_debug2(session, WT_VERB_CHECKPOINT_CLEANUP,
```

Comment:
The log message is saying we read into the cache because of obsolete time window. But this flow can be possible to hit when the page have stop durable timestamp also.

Reply:
That's true, my bad. Fixed in 846cae72b87c56da8acdae2c1d23489d3edf91df.